File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int n=nums.size();
        unordered_map<int,int>mpp;
        mpp[0]=1;
        int prefixSum=0,ans=0;
        for(int i=0;i<n;i++){
            prefixSum=(prefixSum+nums[i]%k+k)%k;
            ans+=mpp[prefixSum];
            mpp[prefixSum]++;
            
        }
        return ans;
        
    }
};

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int ans=0,prefsum=0,n=nums.length;
        map.put(0,1);
        for(int i=0;i<n;i++){
            prefsum=(prefsum+nums[i]%k+k)%k;
             if (map.containsKey(prefsum)) {
                ans += map.get(prefsum);
                map.put(prefsum, map.get(prefsum) + 1);
            } else {
                map.put(prefsum, 1);
            }
        }
        return ans;
    }
}